File: NeuralNet/neuralnet.py
import numpy as np


class NeuralNet:
    def __init__(self, num_features: int, hidden_layer_size: int, categories: int):
        self.W1 = np.random.rand(hidden_layer_size, num_features) - 0.5
        self.b1 = np.random.rand(hidden_layer_size, 1) - 0.5
        self.W2 = np.random.rand(categories, hidden_layer_size) - 0.5
        self.b2 = np.random.rand(categories, 1) - 0.5

        self.Z1 = np.random.rand(hidden_layer_size, 1)
        self.A1 = np.random.rand(hidden_layer_size, 1)
        self.Z2 = np.random.rand(categories, 1)
        self.A2 = np.random.rand(categories, 1)

    @staticmethod
    def ReLU(Z: np.ndarray) -> np.ndarray:
        return np.maximum(Z, 0)

    @staticmethod
    def softmax(Z: np.ndarray) -> np.ndarray:
        e = np.exp(Z - np.max(Z, 0))
        s = np.sum(e, 0)
        return e / s

    def forward_prop(self, X: np.ndarray) -> np.ndarray:
        self.Z1 = self.W1.dot(X) + self.b1
        # W1 10,784 ||| X 784,60000 ||| W.X 10,60000

        self.A1 = NeuralNet.ReLU(self.Z1)
        self.Z2 = self.W2.dot(self.A1) + self.b2
        self.A2 = NeuralNet.softmax(self.Z2)
        return self.A2

    @staticmethod
    def one_hot(Y: np.ndarray) -> np.ndarray:
        one_hot_Y = np.zeros((Y.size, Y.max() + 1))  # create matrix with correct size
        # np.arange(Y.size) - will return a list from 0 to Y.size
        # Y - will contain the column index to set the value of 1
        one_hot_Y[np.arange(Y.size), Y] = 1
        return one_hot_Y.T  # transpose

    @staticmethod
    def deriv_ReLU(Z: np.ndarray) -> np.ndarray:
        return Z > 0

    def back_prop(
        self,
        X: np.ndarray,
        one_hot_Y: np.ndarray,
        alpha: float,
    ) -> None:
        m = one_hot_Y.shape[1]
        dZ2 = self.A2 - one_hot_Y
        dW2 = 1 / m * dZ2.dot(self.A1.T)
        db2 = 1 / m * np.sum(dZ2, 1).reshape(self.b2.shape)
        dZ1: np.ndarray = self.W2.T.dot(dZ2) * NeuralNet.deriv_ReLU(self.Z1)
        dW1 = 1 / m * dZ1.dot(X)
        db1 = 1 / m * np.sum(dZ1, 1).reshape(self.b1.shape)

        self.W1 = self.W1 - alpha * dW1
        self.b1 = self.b1 - alpha * db1
        self.W2 = self.W2 - alpha * dW2
        self.b2 = self.b2 - alpha * db2

    @staticmethod
    def get_predictions(A2: np.ndarray) -> np.ndarray:
        return np.argmax(A2, 0)

    @staticmethod
    def get_correct_prediction(predictions: np.ndarray, Y: np.ndarray) -> int:
        return np.sum(predictions == Y)

    @staticmethod
    def get_accuracy(correct_prediction: int, size: int):
        return 1.0 * correct_prediction / size

File: src/NeuralNet/NeuralNet.cpp
#include <NeuralNet/NeuralNet.hpp>
#include <unsupported/Eigen/MatrixFunctions>

NeuralNet::NeuralNet(const int num_features, const int hidden_layer_size, const int categories)
{
    // Random generates [-1:1]. Numpy is [0:1]
    W1 = Eigen::MatrixXf::Random(hidden_layer_size, num_features);
    W1 = W1.array() / 2.0f;
    b1 = Eigen::VectorXf::Random(hidden_layer_size);
    b1 = b1.array() / 2.0f;
    W2 = Eigen::MatrixXf::Random(categories, hidden_layer_size);
    W2 = W2.array() / 2.0f;
    b2 = Eigen::VectorXf::Random(categories, 1);
    b2 = b2.array() / 2.0f;
}

Eigen::MatrixXf NeuralNet::ReLU(const Eigen::MatrixXf &Z)
{
    return Z.cwiseMax(0);
}

Eigen::MatrixXf NeuralNet::Softmax(Eigen::MatrixXf &Z)
{
    Eigen::MatrixXf e = Z.array().exp();
    Eigen::MatrixXf s = e.colwise().sum();
    for (int c = 0; c < e.cols(); c++)
    {
        e.col(c) = e.col(c) / s(c);
    }
    return e;
}

Eigen::MatrixXf NeuralNet::forward_prop(const Eigen::MatrixXf &X)
{
    Z1 = W1 * X;
    for (int c = 0; c < Z1.cols(); c++)
    {
        Z1.col(c) = Z1.col(c) + b1;
    }
    A1 = ReLU(Z1);

    Z2 = W2 * A1;
    for (int c = 0; c < Z2.cols(); c++)
    {
        Z2.col(c) = Z2.col(c) + b2;
    }
    A2 = Softmax(Z2);

    return A2;
}

Eigen::MatrixXf NeuralNet::one_hot_encode(Eigen::VectorXf &Z)
{
    Eigen::MatrixXf o = Eigen::MatrixXf::Zero(Z.rows(), Z.maxCoeff() + 1);

    for (int r = 0; r < Z.rows() - 1; r++)
    {
        o(r, int(Z(r))) = 1;
    }
    return o.transpose();
}

Eigen::MatrixXf NeuralNet::deriv_ReLU(Eigen::MatrixXf &Z)
{
    const Eigen::Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> b2 = (Z.array() > 0);
    return b2.unaryExpr([](const bool x) { return x ? 1.0f : 0.0f; });
}

void NeuralNet::back_prop(
        const Eigen::MatrixXf &X, const Eigen::VectorXf &Y, const Eigen::MatrixXf &one_hot_Y,
        const float alpha)
{
    const int y_size = Y.rows();

    const Eigen::MatrixXf dZ2 = A2 - one_hot_Y;
    dW2 = dZ2 * A1.transpose() / y_size;
    db2 = dZ2.sum() / y_size;

    const Eigen::MatrixXf dZ1 = (W2.transpose() * dZ2).cwiseProduct(deriv_ReLU(Z1));
    dW1 = dZ1 * X.transpose() / y_size;
    db1 = dZ1.sum() / y_size;

    W1 = W1 - dW1 * alpha;
    b1 = b1.array() - db1 * alpha;
    W2 = W2 - dW2 * alpha;
    b2 = b2.array() - db2 * alpha;
}

Eigen::VectorXf NeuralNet::get_predictions(Eigen::MatrixXf &P)
{
    Eigen::VectorXf p = Eigen::VectorXf::Zero(P.cols()).array() - 1;
    Eigen::Index maxIndex;
    for (int c = 0; c < P.cols(); c++)
    {
        P.col(c).maxCoeff(&maxIndex);
        p(c) = maxIndex;
    }
    return p;
}

int NeuralNet::get_correct_prediction(const Eigen::VectorXf &p, const Eigen::VectorXf &y)
{
    const Eigen::Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> e = p.cwiseEqual(y);
    const Eigen::VectorXi e_int = e.unaryExpr([](const bool x) { return x ? 1 : 0; });
    return e_int.sum();
}

float NeuralNet::get_accuracy(const int correct_prediction, const int size)
{
    return 1.0f * correct_prediction / size;
}
